portfolioGeneration.py

```diff
@@ -0,0 +1,166 @@
+        cov_matrix (np.ndarray): Covariance matrix of asset returns.
+        annualized (bool): Whether to annualize the volatility (default is False).
+
+    Returns:
+        portfolio_volatility (float): Portfolio volatility (risk).
+    """
+    if annualized:
+        portfolio_volatility = np.sqrt(calculate_portfolio_covariance(weights, cov_matrix) * 252) # annualized to 252 trading days
+    else:
+        portfolio_volatility = np.sqrt(calculate_portfolio_covariance(weights, cov_matrix))
+   
+    return portfolio_volatility
+
+
+def generate_and_plot_random_portfolios(returns, cov_matrix, num_portfolios=10000, num_assets=None):
+    """
+    Generate and plot random portfolios on the efficient frontier.
+
+    Parameters:
+        returns (np.ndarray): Array of expected returns for each asset.
+        cov_matrix (np.ndarray): Covariance matrix of asset returns.
+        num_portfolios (int): Number of random portfolios to generate (default is 10000).
+        num_assets (int): Number of assets to consider in each portfolio (default is None).
+
+    Returns:
+        None
+    """
+    portfolio_returns = []
+    portfolio_volatilities = []
+
+    # If num_assets is not specified, use all assets
+    if num_assets is None:
+        num_assets = len(returns)
+
+    for _ in range(num_portfolios):
+        
+        # Create a vector of weights for the entire portfolio
+        full_weights = np.zeros(len(returns)) 
+
+        # Generate random weights that sum to 1 for the specified number of assets
+        weights = generate_dirichlet_portfolios(num_assets)
+
+        # Randomly select the positions (assets) for the held assets
+        asset_positions = random.sample(range(len(returns)), num_assets)
+
+        # Set the weights for the held assets
+        full_weights[asset_positions] = weights
+
+        # Normalize the weights so they sum to 1
+        full_weights /= full_weights.sum()
+
+        # Calculate the portfolio return and volatility
+        portfolio_return = calculate_portfolio_return(full_weights, returns)
+        portfolio_volatility = calculate_portfolio_volatility(full_weights, cov_matrix, annualized=True)
+
+        portfolio_returns.append(portfolio_return)
+        portfolio_volatilities.append(portfolio_volatility)
+
+    # Convert the lists to NumPy arrays
+    portfolio_returns = np.array(portfolio_returns)
+    portfolio_volatilities = np.array(portfolio_volatilities)
+
+    # Plot the efficient frontier
+    plt.figure(figsize=(10, 6))
+    plt.scatter(
+        portfolio_volatilities,
+        portfolio_returns,
+        c=portfolio_returns / portfolio_volatilities,
+        marker="o",
+    )
+    plt.title("Efficient Frontier")
+    plt.xlabel("Volatility (Risk)")
+    plt.ylabel("Return")
+    plt.colorbar(label="Sharpe Ratio")
+    plt.grid(True)
+    plt.show()
+
+
+# Run the code to generate and plot random portfolios
+if __name__ == "__main__":
```

Comment:
looks good to me. since we have so many random generating functions, we could abstract them out into some utils library later on.